devlib/collector/perfetto.py

```diff
@@ -0,0 +1,106 @@
+#    Copyright 2023 ARM Limited
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+import os
+import subprocess
+from shlex import quote
+
+from devlib.host import PACKAGE_BIN_DIRECTORY
+from devlib.collector import (CollectorBase, CollectorOutput,
+                              CollectorOutputEntry)
+from devlib.exception import TargetStableError
+
+OUTPUT_PERFETTO_TRACE = 'trace.perfetto-trace'
+
+
+class PerfettoCollector(CollectorBase):
+    """
+    Perfetto is a production-grade open-source stack for performance instrumentation
+    and trace analysis developed by Google. It offers services and libraries for
+    recording system-level and app-level traces, native + java heap profiling,
+    a library for analyzing traces using SQL and a web-based UI to visualize and
+    explore multi-GB traces.
+
+    This collector takes a path to a perfetto config file saved on disk and passes
+    it directly to the tool.
+
+    On Android platfroms Perfetto is included in the framework starting with Android 9.
+    On Android 8 and below, follow the Linux instructions below to build and include
+    the standalone tracebox binary.
+
+    On Linux platforms, either traced (Perfetto tracing daemon) needs to be running
+    in the background or the tracebox binary needs to be built from source and placed
+    in the Package Bin directory. The build instructions can be found here:
+
+    https://perfetto.dev/docs/contributing/build-instructions
+
+    After building the 'tracebox' binary should be copied to devlib/bin/<arch>/.
+
+    For more information consult the official documentation:
+    https://perfetto.dev/docs/
+    """
+
+    def __init__(self, target, config=None):
+        super().__init__(target)
+        self.target_output_file = target.path.join(self.target.working_directory, OUTPUT_PERFETTO_TRACE)
+        self.bg_cmd = None
+        self.config = config
+
+        os_version = target.os_version['release']
+        self.target_binary = 'perfetto'
+        if target.os == 'android' and os_version > 8:
+            # Android 9 and 10 require traced to be enabled manually
+            if os_version == '9' or os_version == '10':
+                target.execute('setprop persist.traced.enable 1')
+        elif target.os in ['linux', 'android'] and not target.is_running('traced'):
```

Comment:
Hmm that's intentional though. We only check for `traced` so that we can use `tracebox` if it's missing. If `traced` is present this means that we don't need to use `tracebox` and should call `perfetto` instead.
Calling `tracebox` with `traced` running in the background would be an odd halfway setup I think. From their docs:

> Due to Perfetto's [service-based architecture](https://perfetto.dev/docs/concepts/service-model), in order to capture a trace, the traced (session daemon) and traced_probes (probes and ftrace-interop daemon) need to be running. As per Perfetto v16, the tracebox binary bundles together all the binaries you need in a single executable (a bit like toybox or busybox).